### PR TITLE
Use type inference instead of test_value

### LINKED ISSUE #32: why use typeof(f(testvalue(data)))?
Current, to determine the `eltype` of a `MappedArray`, this package computes `typeof(f(testvalue(data)))`. While it's commendable to try to avoid using `Core.Compiler.return_type` for this, I think it might be misguided. Not only could this be incorrect for heterogeneously typed arrays, but it's potentially very slow.

I'm very tempted to think this is a case where reaching into internals to just ask the compiler for the return type is justified.

### PR DESCRIPTION
**Pull request by MasonProtter**
closes #32

### CODE REVIEW

File: src/MappedArrays.jl [hunk 1]
```diff
@@ -58,7 +58,7 @@ not set them).
 When multiple input arrays are supplied, `M[i] = f(A[i], B[i], C[i]...)`.
 """
 function mappedarray(f, data::AbstractArray)
-    T = typeof(f(testvalue(data)))
+    T = Core.Compiler.return_type(f, (eltype(data),))
```
Comment (timholy) [inline src/MappedArrays.jl:61]: There is no `Base.return_type` :smile:

Comment (timholy) [inline src/MappedArrays.jl:61]: Not inferrable (see the test failure).

**Comment by timholy** [comment]
You want to use `Base._return_type(f, Tuple{eltype(data)})`.

**Comment by MasonProtter** [comment]
```julia
julia> Base._return_type === Core.Compiler.return_type
true
```
If you prefer `_return_type` for stylistic reasons we can use it though.

File: src/MappedArrays.jl (hunk 1 above)
Comment (MasonProtter) [reply src/MappedArrays.jl:61]: That's interesting, it's passing for me locally on 1.5.2, I wonder what could be causing that.

Comment (MasonProtter) [reply src/MappedArrays.jl:61]: Could it be that travis is activating some flags to reduce inference?

Comment (MasonProtter) [reply src/MappedArrays.jl]: Nevermind, somehow the tests were happening on the wrong branch,

File: Project.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 name = "MappedArrays"
 uuid = "dbb5928d-eab1-5f90-85c2-b9b0edb7c900"
-version = "0.3.0"
+version = "0.4.0"
```
Comment (johnnychen94) [inline Project.toml:3]: LOL we just released v0.3.0 yesterday

**Review by johnnychen94 — COMMENTED** [review]
Looks good but tests are needed.

**Comment by MasonProtter** [comment]
I'm not really sure what else to test since the test suite itself already tests for concrete inference.

**Comment by johnnychen94** [comment]
Maybe

```julia
mappedarray(+, [1, missing, 3], [missing, 5, 6.0])

_zero(x) = x > 0 ? x : 0
mappedarray(_zero, [1.0, 1])

of_eltype(Union{Missing, Float64}, [1, 2])
```

File: Project.toml (hunk 2 above)
Comment (johnnychen94) [reply Project.toml:3]: Unfortunately, we definitely need 0.4 here.

**Comment by cscherrer** [comment]
Any update on this? This PR also seems to remove the need for https://github.com/JuliaArrays/MappedArrays.jl/pull/41

Some of my work is blocked until one of these can go through, so I'd love for it to be able to move forward.

**Comment by timholy** [comment]
Best to add the various test cases (esp. https://github.com/JuliaArrays/MappedArrays.jl/pull/34#issuecomment-706265722) and then we can merge.

**Comment by timholy** [comment]
In other words, it's lonely and needs a friend to help it cross the finish line. You could be the one!

**Comment by johnnychen94** [comment]
closed in favor of #43

**Pull request closed without merging**